File: packages/rest/src/media-type.test.ts
import { describe, expect, it } from 'vitest';

import { normalizeMediaType } from './media-type.js';

describe('normalizeMediaType', () => {
  it.each([
    ['images', 'IMAGE'],
    ['videos', 'VIDEO'],
    ['audio', 'AUDIO'],
    ['documents', 'DOCUMENT'],
  ] as const)('maps the S3 folder %s to %s', (folder, expected) => {
    // These four are the only values s3-client's getMediaFolder can produce
    // (s3-client.ts:13-32, 38-44).
    expect(normalizeMediaType(folder)).toBe(expected);
  });

  it.each(['IMAGE', 'VIDEO', 'AUDIO', 'DOCUMENT'] as const)(
    'passes an already-correct %s through unchanged',
    (name) => {
      expect(normalizeMediaType(name)).toBe(name);
    },
  );

  it.each([
    ['an unrecognized string', 'sticker'],
    ['an empty string', ''],
    ['undefined', undefined],
    ['null', null],
    ['a number', 4],
    ['an object', { mediaType: 'images' }],
  ])('falls back to DOCUMENT for %s', (_label, value) => {
    // Mirrors s3-client.ts:43's own fallback; core then degrades DOCUMENT to
    // FILE, so an unknown kind still sends rather than failing the upload.
    expect(normalizeMediaType(value)).toBe('DOCUMENT');
  });

  it('tolerates surrounding whitespace and mixed case', () => {
    expect(normalizeMediaType('  Images ')).toBe('IMAGE');
  });
});

File: packages/dart_rest/test/media_type_test.dart
/// Reproduces `packages/rest/src/media-type.test.ts`.
library;

import 'package:dhaam_chat_rest/src/media_type.dart';
import 'package:test/test.dart';

void main() {
  group('normalizeMediaType', () {
    for (final (String folder, String expected) in <(String, String)>[
      ('images', 'IMAGE'),
      ('videos', 'VIDEO'),
      ('audio', 'AUDIO'),
      ('documents', 'DOCUMENT'),
    ]) {
      test('maps the S3 folder $folder to $expected', () {
        // These four are the only values s3-client's getMediaFolder can
        // produce.
        expect(normalizeMediaType(folder), expected);
      });
    }

    for (final String name in <String>['IMAGE', 'VIDEO', 'AUDIO', 'DOCUMENT']) {
      test('passes an already-correct $name through unchanged', () {
        expect(normalizeMediaType(name), name);
      });
    }

    for (final (String label, Object? value) in <(String, Object?)>[
      ('an unrecognized string', 'sticker'),
      ('an empty string', ''),
      ('null', null),
      ('a number', 4),
      ('an object', <String, Object?>{'mediaType': 'images'}),
      ('a list', <Object?>['images']),
      ('a bool', true),
    ]) {
      test('falls back to DOCUMENT for $label', () {
        // Mirrors s3-client's own fallback; DOCUMENT then degrades to a
        // generic FILE attachment, so an unknown kind still sends rather than
        // failing the upload. An upload is not worth failing over a label.
        expect(normalizeMediaType(value), 'DOCUMENT');
      });
    }

    test('tolerates surrounding whitespace and mixed case', () {
      expect(normalizeMediaType('  Images '), 'IMAGE');
    });
  });
}
